### PR TITLE
thrift-proxy: add support for cluster header based routing

### PR DESCRIPTION
**Pull request by rgs1**
Fixes: #8991

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

File: source/extensions/filters/network/thrift_proxy/router/router_impl.cc [hunk 1]
```diff
@@ -211,24 +212,25 @@ FilterStatus Router::messageBegin(MessageMetadataSharedPtr metadata) {
 
   route_entry_ = route_->routeEntry();
 
-  Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(route_entry_->clusterName());
+  const std::string& cluster_name = getClusterName(metadata);
```
Comment (zuercher) [inline source/extensions/filters/network/thrift_proxy/router/router_impl.cc]: The cluster lookup is also performed in the ratelimit filter (and potentially custom filters).

The intention of the design is that `RouteEntry::clusterName()` returns the selected cluster name no matter how it was chosen. So I think the correct way to implement this is to introduce a new class, similar to `WeightedClusterEntry`, constructed with a parent `RouteEntryImplBase*` and a cluster name. (This all works as in the HTTP router where the analogous class is called `DynamicRouteEntry`). 

In `RouteEntryImplBase::clusterEntry` you can then detect the cluster_header config and extract the chosen cluster name and return a newly constructed `DynamicRouteEntry` (you'll have to plumb through the metadata, but I think it's available at all the clusterEntry call sites).

Then this line revert to calling `route_entry_->clusterName()` and any other calls to `RouteEntry::clusterName()` will get the correct cluster name. (It's fine to cache the name in this method for logging.)

Comment (rgs1) [reply source/extensions/filters/network/thrift_proxy/router/router_impl.cc]: Ah -- I see. Thanks for the quick review and the detailed explanation, I'll rework this to follow that model.

**Comment by rgs1** [comment]
cc: @htuch for API.

File: api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto [hunk 2]
```diff
@@ -82,6 +82,12 @@ message RouteAction {
     // request is routed to one of the upstream clusters based on weights
     // assigned to each cluster.
     WeightedCluster weighted_clusters = 2;
+
+    // Envoy will determine the cluster to route to by reading the value of the
+    // Thrift header named by cluster_header from the request headers. If the
+    // header is not found or the referenced cluster does not exist, Envoy will
+    // return an exception.
```
Comment (htuch) [inline api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: What does returning an exception mean from an API perspective? This seems alarming behavior 😨

Comment (rgs1) [reply api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: In thrift land it's the equivalent of a 503... I tried to be consistent with:

https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/route/route.proto#L670

Shall I say a "thrift exception" ?

Comment (rgs1) [reply api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: Or even better, a "thrift response exception"?

Comment (rgs1) [reply api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: Improved the wording a bit.

Comment (htuch) [reply api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto]: I see, Thrift has a notion of exception messages. LGTM, thanks.

**Pull request merged**